internal/utils/ctrlcreader.go

```diff
@@ -0,0 +1,69 @@
+// Copyright 2020-2021 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
```

Comment:
This looks great! The Close() vs. ctx cancel difference is nice. :+1:  The function docs might need another pass to reflect this, it wasn't immediately clear to me that there were 2 ways of making Read return, and that they exhibit different behaviours.